### ISSUE TITLE
Feat: Performance tests

### ISSUE THREAD

**Issue by YoniSchirris** [lucidrains/byol-pytorch#3]
First of all: Many thanks for this rapid implementation of the paper's method. The implementation looks very clean.

I was wondering if you've done any performance testing to see if the implementation works as expected. E.g. by reproducing the Linear Evaluation on ImageNet experiment with BYOL.

**Comment by lucidrains** [comment]
Thanks! I'll get around to this once my GPUs free up!

**Comment by Franklin-Yao** [comment]
Anyone verified the correctness of the code?

**Comment by YoniSchirris** [comment]
Even though the loss increases after a few epochs for my task, it achieves a similar downstream task performance as a SimCLR implementation I'm using for histopathology research.

**Comment by pribadihcr** [comment]
@YoniSchirris, How many data and epochs to achieve the expected result? thanks

**Comment by YoniSchirris** [comment]
Implementation and performance seems fine. 

Data / epochs / model is too domain-specific to be meaningful, but I used 45 epochs of random init shufflenetv2 on millions on histopathology images at 1.1 microns per pixel of 224x224 px. Achieved similar performance to another SimCLR v1 implementation, which was better than an Imagenet pretrained network.

Haven't done any formal tests on any standardized datasets, though.

**Comment by binli123** [comment]
> Implementation and performance seems fine.
> 
> Data / epochs / model is too domain-specific to be meaningful, but I used 45 epochs of random init shufflenetv2 on millions on histopathology images at 1.1 microns per pixel of 224x224 px. Achieved similar performance to another SimCLR v1 implementation, which was better than an Imagenet pretrained network.
> 
> Haven't done any formal tests on any standardized datasets, though.

Have you done any controlled experiments and comparisons between SimCLR and BYOL on WSI datasets? i.e., same backbone and same batch size? I used ResNet18 as the backbone and the loss didn't drop much but started to rise after several epochs. Using the pre-trained model to compute the embedders leading to significantly worse performance in my downstream classification task compared to SimCLR, yet much better than ImageNet.

**Comment by binli123** [comment]
> > > Implementation and performance seems fine.
> > > Data / epochs / model is too domain-specific to be meaningful, but I used 45 epochs of random init shufflenetv2 on millions on histopathology images at 1.1 microns per pixel of 224x224 px. Achieved similar performance to another SimCLR v1 implementation, which was better than an Imagenet pretrained network.
> > > Haven't done any formal tests on any standardized datasets, though.
> > 
> > 
> > Have you done any controlled experiments and comparisons between SimCLR and BYOL on WSI datasets? i.e., same backbone and same batch size? I used ResNet18 as the backbone and the loss didn't drop much but started to rise after several epochs. Using the pre-trained model to compute the embedders leading to significantly worse performance in my downstream classification task compared to SimCLR, yet much better than ImageNet.
> 
> Same model, yes. Same batch size, no; that doesn't makes sense since these models very differently depend on the batch size. For me, performance was similar for both, which was much better than ImageNet. Loss of BYOL indeed increased, which I expect to happen due to the dynamics of the two networks (target network slowly catches up, making the task much harder after a few epochs)

In terms of choosing the method to pre-train the model given the available GPU resources, a comparison with the same training memory requirement and training time would be useful. Besides, I also noticed in the [paper](https://arxiv.org/pdf/2006.07733.pdf) Figure 6 (appendix), their upper bound performance (fully supervised ones) was significantly better than the ones reported in SimCLR. So I am skeptical about how better it actually is compared to SimCLR or MoCo v2.
